extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClients.java

```diff
@@ -299,6 +269,39 @@ private MongoClientSettings createMongoConfiguration(MongoClientConfig config) {
         return settings.build();
     }
 
+    private void configureCodecRegistry(CodecRegistry defaultCodecRegistry, MongoClientSettings.Builder settings) {
+        List<CodecProvider> providers = new ArrayList<>();
+        if (!mongoClientSupport.getCodecProviders().isEmpty()) {
+            providers.addAll(getCodecProviders(mongoClientSupport.getCodecProviders()));
+        }
+        // add pojo codec provider with automatic capabilities
+        // it always needs to be the last codec provided
+        PojoCodecProvider.Builder pojoCodecProviderBuilder = PojoCodecProvider.builder()
+                .automatic(true)
+                .conventions(Conventions.DEFAULT_CONVENTIONS);
+        // register bson discriminators
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        for (String bsonDiscriminator : mongoClientSupport.getBsonDiscriminators()) {
+            try {
+                pojoCodecProviderBuilder
+                        .register(ClassModel.builder(Class.forName(bsonDiscriminator, true, classLoader))
+                                .enableDiscriminator(true).build());
+            } catch (ClassNotFoundException e) {
+                // Ignore
+            }
+        }
+        // register property codec provider
+        if (!mongoClientSupport.getPropertyCodecProviders().isEmpty()) {
+            pojoCodecProviderBuilder.register(getPropertyCodecProviders(mongoClientSupport.getPropertyCodecProviders())
+                    .toArray(new PropertyCodecProvider[0]));
+        }
+        CodecRegistry registry = !providers.isEmpty() ?
+                fromRegistries(fromProviders(providers), defaultCodecRegistry,
+                        fromProviders(pojoCodecProviderBuilder.build())) :
+                fromRegistries(defaultCodecRegistry, fromProviders(pojoCodecProviderBuilder.build()));
```

Comment:
is it correct? providers provided by `pojoCodecProviderBuilder` will be registered  after `defaultCodecRegistry`- this will be the same situation as it was with custom providers :(

Reply:
In theory, no.  Custom defined codecs, as I understand how this is all formulated, are in the `providers` list where POJO codecs are registered with `pojoCodecProviderBuilder`.  So the codecs for entities will be resolved last, yes, but in practice this will only be a problem if the `defaultCodecRegistry` somehow has a codec for your entity types which seems a bit of a stretch.  But should you wish to provide your own codec for a non-entity type, this should be sufficient.